### ISSUE TITLE
rule request: warn on template literal syntax inside normal strings

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#5850]
**When does this rule warn? Please describe and show example code:**
Any time a string literal with single or double quotes has `${foo}` in it, *where `foo` would evaluate in the scope of the string*.

```js
const foo = 3;

// warns
const bar = 'foo: ${foo}';
const baz = "foo: ${foo}";

// does not warn
const bar = `foo: ${foo}`;
const baz = `foo: ${foo}`;
const quux = 'nonexistent: ${nonexistent}';
const xyzzy = "nonexistent: ${nonexistent}";
```

**Is this rule preventing an error or is it stylistic?**
This rule is attempting to prevent a developer error from using the wrong kind of quotes, when they intended to use a template literal.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**
I think this might be a very common footgun for developers new to ES6, or migrating codebases to it. For example, while complying with the `prefer-template` rule, they may consolidate the string, but forget to use the proper quotes. Alternatively, a well-meaning developer may not know what template literals are, and may "helpfully" change them to single/double quotes.

**Are you willing to create the rule yourself?**
I lack the expertise, but would give it a shot when I can find the time, if nobody else is willing.

**Comment by platinumazure** [comment]
Hmm... Checking for a regular expression pattern `/\$\{[^}]+\}/` would be easy enough, but adding in the evaluation bit could be harder. Do you see that as critical to this rule, or do you think a heuristic would be good enough?

Bear in mind that people can choose not to use the new rule if they are commonly using `${something}` in non-template string literals.

**Comment by ljharb** [comment]
I think that the rule would be much more valuable that way - but, I think the regex approach would be a great starting point. It could always be improved in a non-major version bump.

**Comment by nzakas** [comment]
I don't think there's enough evidence that this is a real-world problem to include this in the core right now. I'll leave this open in case anyone believes this reaches the level of "must-have" in ESLint.

**Comment by ljharb** [comment]
This came up as an actual real-world problem, in user code, which is why I proposed the rule. The user was confused why our eslint config was warning on an unused variable. We figured out that it was the wrong quote type, but had they used the variable elsewhere in the function, they might never have discovered it.

**Comment by platinumazure** [comment]
Given the real-world problem here, I would say this is very situational, but (when the situation can come up) highly compelling. I honestly think this is worth another look.

That said, @ljharb, even if this doesn't make core, you are more than welcome to create this as a custom rule or as part of a plugin. If I were you, I would find a plugin that focuses on template strings and propose this as an enhancement there, in case it's not accepted into core.

**Comment by ljharb** [comment]
@platinumazure the only one i was able to find is already included in core. obviously if it's not accepted into core, I'll try to create a plugin, but I'd really like to see it included.

**Comment by nzakas** [comment]
Sorry, what I meant is, I generally feel that we need evidence that the problem has some level of ubiquity before we accept something as a core rule. The use case of, "it happened once to us and I want to prevent it from happening again" is precisely the use case for custom rules (literally this is what caused me to create ESLint).

I'll leave this open for a bit to see if anyone on the team wants to disagree.

**Comment by ljharb** [comment]
@nzakas what's the path for custom plugins to be included in core, without having to copy and paste the code over?

**Comment by nzakas** [comment]
@ljharb I'm not quite sure what you're asking. Plugins are never included in the core, and we rarely migrate a rule from a plugin into the core (because if it exists in a plugin, why not keep it there?). If we decide something is so important that it should be in the core, in the past we have just copied it over.

So, it looks like no one else felt strongly enough for this rule to champion it, so a custom rule is your best bet at this point.

**Comment by ljharb** [comment]
That policy means I'm strongly disincentivized to ever create a custom plugin, because by doing so, I've doomed my rule to never be included in core.

**Comment by platinumazure** [comment]
@ljharb Did you end up creating a custom rule?

@nzakas Given the second issue we've had on this, is it worth reopening this for consideration for core?

**Comment by ljharb** [comment]
@platinumazure no, creating a custom rule for a generic warning seems silly when it isn't a path to it being included in core, so I haven't made the time yet. I'd certainly use one if it were created tho.
